add support for repacking a variant

jpculp opened the pull request:
**Description of changes:**

This adds a new tool to take an existing Bottlerocket image and repackage it with updated signatures and artifacts.

**Testing done:**

- [x] Repack `aws-k8s-1.27` _(no secure boot)_.
- [x] Repack `aws-k8s-1.28` _(secure boot)_.
- [x] Repack `vmware-k8s-1.28`.
- [x] Repack `aws-k8s-1.28` with custom CA bundle using #232.
- [x] Verify nothing prints on successful repack.
- [x] Verify everything prints on unsuccessful repack.
- [x] Verify logs are wiped between repacks.
- [x] Verify log symlink is correct.
- [x] Verify `clean-logs` cleans the logs.
- [x] Verify `exit 1` if something goes wrong in `debugfs`.
 
**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+# Process and stage the input images to the working directory.
+stage_images "${INPUT_DIR}" "${OUTPUT_FMT}" OS_IMAGE DATA_IMAGE
+
+# Validate that our input image matches the partition plan.
+SGDISK_OUTPUT="$(sgdisk --print "${OS_IMAGE}")"
+for part in EFI-A BOOT-A ROOT-A HASH-A; do
+  case "${part}" in
+  EFI-A)
+    sgdisk_start="$(awk /EFI-SYSTEM/'{print $2}' <<<"${SGDISK_OUTPUT}")"
+    sgdisk_size="$(awk /EFI-SYSTEM/'{print $4}' <<<"${SGDISK_OUTPUT}")"
+    ;;
+  *)
+    sgdisk_start="$(awk /BOTTLEROCKET-"${part}"/'{print $2}' \
+      <<<"${SGDISK_OUTPUT}")"
+    sgdisk_size="$(awk /BOTTLEROCKET-"${part}"/'{print $4}' \
+      <<<"${SGDISK_OUTPUT}")"
+    ;;
+  esac
+
+  # 1 MiB contains 2048 512-byte sectors.
+  sgdisk_start_mib="$((sgdisk_start / 2048))"
+
+  if [[ "${partoff["${part}"]}" -ne "${sgdisk_start_mib}" ]]; then
+    echo "start mismatch between partition plan and disk: '${part}'" >&2
+    exit 1
+  fi
+
+  if [[ "${partsize["${part}"]}" -ne "${sgdisk_size%.0}" ]]; then
+    echo "size mismatch between partition plan and disk: '${part}'" >&2
+    exit 1
+  fi
+done
+
```
I'm probably going to replace this block with a new `get_partition_sizes` function in `partyplanner` and compare one array to the other.

jpculp replied:
Replaced.

jpculp commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+###############################################################################
+# Section 3: replace needed artifacts
+
+# TODO - CA bundle code goes here
```
This might be a separate PR.

jpculp replied:
If this merges first I'll add a commit to #232.

jpculp commented on twoliter/embedded/imghelper:
```diff
@@ -244,6 +282,18 @@ mkfs_data_xfs() {
   dd if="${bottlerocket_data}" of="${target}" conv=notrunc bs=1M seek="${offset}"
 }
 
+check_image_size() {
```
I considered putting this in `partyplanner`, but if it lives here then we can use it in `generate_verity_root` without sourcing `partyplanner`.

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+OUTPUT_DIR="${OUTPUT_DIR}/${VERSION_ID}-${BUILD_ID}"
+mkdir -p "${OUTPUT_DIR}"
+
+declare OS_IMAGE DATA_IMAGE
```
nit: declare immediately before use by `stage_images`

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+  "${OS_IMAGE_SIZE_GIB}" "${DATA_IMAGE_SIZE_GIB}" "${PARTITION_PLAN}" \
+  partsize partoff
+
+for part in EFI-A BOOT-A ROOT-A HASH-A; do
+  # Each partition is aligned to a 1 MiB boundary, and extends to the sector
+  # before the next partition starts. Specify the end point in sectors so we
+  # can subtract a sector to fix the off-by-one error that comes from adding
+  # start and size together. (1 MiB contains 2048 512-byte sectors.)
+  part_start="${partoff["${part}"]}"
+  part_end="$((part_start + partsize["${part}"]))"
+  part_end="$((part_end * 2048 - 1))"
+done
```
Can we remove this? `part_end` doesn't seem to be used.

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+
+# Validate and write root image back to the OS image.
+check_image_size "${ROOT_IMAGE}" "${partsize["ROOT-A"]}"
+dd if="${ROOT_IMAGE}" of="${OS_IMAGE}" \
```
We should add `iflag=fullblock` here

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+generate_verity_root "${ROOT_IMAGE}" "${VERITY_IMAGE}" \
+  "${partsize["HASH-A"]}" DM_VERITY_ROOT
+
+# Validate and write root verity back to the OS image. The image size check
+# isn't needed here as it's already done in `generate_verity_root`.
+dd if="${VERITY_IMAGE}" of="${OS_IMAGE}" \
+  conv=notrunc bs=1M seek="${partoff["HASH-A"]}"
```
The comment refers to validation but I just see the write?

jpculp replied:
It's on [L316](https://github.com/bottlerocket-os/twoliter/pull/231/files#diff-0b8c4b6920fcc7cd541e1b6cac7893cf86c9b2e80859c9444ebdf3339b5f3196R316) of `imghelper` since it's checked both in `img2img` and `rpm2img`.

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+
+  # Write the resigned artifacts back to the EFI image.
+  mcopy -i "${EFI_IMAGE}" -ov "${EFI_MOUNT}/EFI/BOOT"/*.efi ::/EFI/BOOT
+  mcopy -i "${EFI_IMAGE}" -ov "${SBKEYS}"/db.{crt,cer} ::/EFI/BOOT
```
Can we call `provide_certs`?

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+# Section 6: update boot partition
+
+GRUB_CONFIG="${BOOT_MOUNT}/grub/grub.cfg"
+[[ -s "${GRUB_CONFIG}" ]]
```
Do we need this check? It seems a bit lonely - we aren't doing similar checks for other files we expect.

jpculp replied:
We can drop.

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,318 @@
+
+popd >/dev/null
+
+# Ensure proper ownership of the final artifacts.
+find "${OUTPUT_DIR}" -type f -print -exec chown 1000:1000 {} \;
```
Might also be worth cleaning up the temp files, as `rpm2img` does.

jpculp replied:
Towards the top of the script there is a trap to clean-up **WORKDIR**, which is where all the temp files are consolidated. Would you prefer it just be at the end like in `rpm2img`?

bcressey replied:
No, the trap is better, I just missed the implicit cleanup.

bcressey reviewed (commented):
Looks good!

jpculp commented:
- Removed unnecessary blocks of code.
- Added `iflag=fullblock` when `dd`ing to the OS image.
- Added clobbering and verbosity to `mcopy` in `provide_certs`.

jpculp committed:
partyplanner: add get_partition_sizes

Populates the caller's tables with sizes and offsets for known
partitions of existing images.

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
Added a more exhaustive comparison of the partition plan and the input image's partition table.

jpculp commented on twoliter/embedded/partyplanner:
```diff
@@ -118,6 +118,94 @@ PRIVATE_SCALE_FACTOR="24"
+      offset="$(awk /EFI-BACKUP/'{print $2}' <<<"${os_image_table}")"
+      sectors="$(awk /EFI-BACKUP/'{print $4}' <<<"${os_image_table}")"
+      ;;
+    RESERVED-A)
+      offset="$(awk /BOTTLEROCKET-RESERV/'{print $2}' \
+      <<<"${os_image_table}" | head -1 || true)"
+      sectors="$(awk /BOTTLEROCKET-RESERV/'{print $4}' \
+      <<<"${os_image_table}" | head -1 || true)"
+      ;;
+    RESERVED-B)
+      offset="$(awk /BOTTLEROCKET-RESERV/'{print $2}' \
+      <<<"${os_image_table}" | tail -1 || true)"
+      sectors="$(awk /BOTTLEROCKET-RESERV/'{print $4}' \
+      <<<"${os_image_table}" | tail -1 || true)"
+      ;;
```
This is a little gross because the labels for **RESERVED-A/B** are being truncated. The alternative is iterating through each partition in the GPT table one-by-one, collecting the detailed information, and populating the arrays that way.

rpkelly commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,302 @@
+
+# Clean up working directory to reduce size of layer.
+cleanup() {
+  [[ -n "${WORKDIR}" ]] && rm -rf "${WORKDIR}"
```
Why not `-d` here?

jpculp commented:
Added logging mechanism to keep successful builds quiet, but still provide context on error.

jpculp committed:
img2img: add support for repacking a variant

This adds a new tool to take an existing Bottlerocket image and
repackage it with updated signatures and artifacts.

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
- Renamed some variables to reflect **p**_(artition)_**t**_(able)_.
- Made `case` indention consistent with the rest of `partyplanner`.
- Switched cleanup check to `-d`.

bcressey commented on twoliter/embedded/Makefile.toml:
```diff
@@ -850,8 +851,15 @@ export PATH="${TWOLITER_TOOLS_DIR}:${PATH}"
+mkdir -p "${OUTPUT_LOGS_DIR}/${BUILDSYS_VERSION_FULL}"
 buildsys repack-variant \
-  --cargo-manifest-dir variants/${BUILDSYS_VARIANT}
+  --cargo-manifest-dir "variants/${BUILDSYS_VARIANT}" \
+  > "${REPACK_OUTPUT_LOG}" || cat "${REPACK_OUTPUT_LOG}"
```
this shows the log on error but clobbers the non-zero exit code, which we need to pass up to avoid reporting success

jpculp commented:
`exit 1` on repack failure en lieu of the return code captured by the conditional.

jpculp committed:
twoliter: log repack-variant output to build/logs

Also adds a 'clean-logs' target, which is part of 'clean'.

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
Added a check to validate `debugfs` commands ran successfully as it does not support return codes _(even when running a single command at a time with `-R`)_.

bcressey commented on twoliter/embedded/img2img:
```diff
@@ -1,3 +1,320 @@
+}
+trap 'cleanup' EXIT
+
+# Since debugfs doesn't pass return codes, extrapolate failures from STDERR.
+check_debugfs_errors() {
```
There are spots in `rpm2img` that can use this check also (can be a follow-up PR).

jpculp replied:
I wasn't aware, but I agree we should check both. I'll try to follow up with a separate PR for that.
